cdc/sink/codec/canal_flat.go

```diff
@@ -0,0 +1,243 @@
+			}
+		}
+	}
+
+	if len(rowData.AfterColumns) > 0 {
+		data = make(map[string]interface{}, len(rowData.AfterColumns))
+		for i := range rowData.AfterColumns {
+			if !rowData.AfterColumns[i].GetIsNull() {
+				data[rowData.AfterColumns[i].Name] = rowData.AfterColumns[i].Value
+			} else {
+				data[rowData.AfterColumns[i].Name] = nil
+			}
+		}
+	} else {
+		// The event type is DELETE
+		// The following line is important because Alibaba's adapter expects this, and so does Flink.
+		data = oldData
+	}
+
+	ret := &canalFlatMessage{
+		ID:            0, // ignored by both Canal Adapter and Flink
+		Schema:        header.SchemaName,
+		Table:         header.TableName,
+		PKNames:       pkNames,
+		IsDDL:         false,
+		EventType:     header.GetEventType().String(),
+		ExecutionTime: header.ExecuteTime,
+		BuildTime:     0, // ignored by both Canal Adapter and Flink
+		Query:         "",
+		SQLType:       sqlType,
+		MySQLType:     mysqlType,
+		Data:          make([]map[string]interface{}, 0),
+		Old:           make([]map[string]interface{}, 0),
+		tikvTs:        e.CommitTs,
+	}
+
+	// We need to ensure that both Data and Old have exactly one element,
+	// even if the element could be nil. Changing this could break Alibaba's adapter
+	ret.Data = append(ret.Data, data)
+	ret.Old = append(ret.Old, oldData)
+
+	return ret, nil
+}
+
+func (c *CanalFlatEventBatchEncoder) newFlatMessageForDDL(e *model.DDLEvent) (*canalFlatMessage, error) {
+	header := c.builder.buildHeader(e.CommitTs, e.TableInfo.Schema, e.TableInfo.Table, convertDdlEventType(e), 1)
+
+	ret := &canalFlatMessage{
+		ID:            0, // ignored by both Canal Adapter and Flink
+		Schema:        header.SchemaName,
+		Table:         header.TableName,
+		IsDDL:         true,
+		EventType:     header.GetEventType().String(),
+		ExecutionTime: header.ExecuteTime,
+		BuildTime:     0, // ignored by both Canal Adapter and Flink
+		Query:         e.Query,
+		tikvTs:        e.CommitTs,
+	}
+	return ret, nil
+}
+
+// EncodeCheckpointEvent is no-op
+func (c *CanalFlatEventBatchEncoder) EncodeCheckpointEvent(ts uint64) (*MQMessage, error) {
+	return nil, nil
+}
+
+// AppendRowChangedEvent implements the interface EventBatchEncoder
+func (c *CanalFlatEventBatchEncoder) AppendRowChangedEvent(e *model.RowChangedEvent) (EncoderResult, error) {
+	msg, err := c.newFlatMessageForDML(e)
+	if err != nil {
+		return EncoderNoOperation, errors.Trace(err)
+	}
+	c.unresolvedBuf = append(c.unresolvedBuf, msg)
+	return EncoderNoOperation, nil
+}
+
+// AppendResolvedEvent receives the latest resolvedTs
+func (c *CanalFlatEventBatchEncoder) AppendResolvedEvent(ts uint64) (EncoderResult, error) {
+	nextIdx := 0
+	for _, msg := range c.unresolvedBuf {
+		if msg.tikvTs <= ts {
+			c.resolvedBuf = append(c.resolvedBuf, msg)
+		} else {
+			break
+		}
+		nextIdx++
+	}
+	c.unresolvedBuf = c.unresolvedBuf[nextIdx:]
+	if len(c.resolvedBuf) > 0 {
+		return EncoderNeedAsyncWrite, nil
+	}
+	return EncoderNoOperation, nil
+}
```

Comment:
```suggestion
func (c *CanalFlatEventBatchEncoder) AppendResolvedEvent(ts uint64) (EncoderResult, error) {
	c.unresolvedIdx := 	i := sort.Search(len(c.unresolvedBuf), func(i int) bool { return c.unresolvedBuf[i].tikvTs > ts })
	if c.unresolvedIdx > 0 {
		return EncoderNeedAsyncWrite, nil
	}
	return EncoderNoOperation, nil
}
```
In `Build`
```go
func (c *CanalFlatEventBatchEncoder) Build() []*MQMessage {
	if c.unresolvedIdx == 0 {
		return nil
	}
	ret := make([]*MQMessage, len(c.unresolvedIdx))
	for i := range ret {
		value, err := json.Marshal(c.unresolvedBuf[i])
		if err != nil {
			log.Fatal("CanalFlatEventBatchEncoder", zap.Error(err))
			return nil
		}
		ret[i] = NewMQMessage(nil, value, c.unresolvedBuf[i].tikvTs)
	}
	c.unresolvedBuf = c.unresolvedBuf[c.unresolvedIdx:]
	return ret
}
```
Maybe this could be more efficient.